File: 0222-Count-Complete-Tree-Nodes/0222.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int: # type: ignore
        if not root: 
            return 0

        left = self.left_height(root)
        right = self.right_height(root)
        
        if left > right:
            return 1 + self.countNodes(root.left) + self.countNodes(root.right)
        else:
            return 2 ** left - 1
        
    def left_height(self, node):
        if not node: 
            return 0
        return 1 + self.left_height(node.left)

    def right_height(self,node):
        if not node: 
            return 0
        return 1 + self.right_height(node.right)

File: 0222-Count-Complete-Tree-Nodes/0222.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        if (root == null)
            return 0;

        int left = leftHeight(root);
        int right = rightHeight(root);

        if (left > right) {
            return 1 + countNodes(root.left) + countNodes(root.right);
        } else {
            return Math.pow(2, left) - 1;
        }

    }

    public int leftHeight(TreeNode node) {
        if (node == null)
            return 0;
        return 1 + leftHeight(node.left);
    }

    public int rightHeight(TreeNode node) {
        if (node == null)
            return 0;
        return 1 + rightHeight(node.right);
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}